ScheduleBlock 2/4 - add Call instruction

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This is 2/4 step of `ScheduleBlock` implementation #5679 


### Details and comments
This PR adds `Call` instruction to pulse instruction set. This instruction is not exposed to end users. However, it implicitly wraps schedules or circuits when `pulse.call` function is used within the pulse builder context. The `Call` instruction explicitly defines a subroutine that is defined outside of the current scope, thus the Qiskit compiler will be able to generate better target code, i.e. the compiler can avoid defining duplicated subprograms multiple times. The logic of the pulse builder is slightly updated to introduce this instruction.

lcapelluto commented on qiskit/pulse/instructions/directives.py:
```diff
@@ -52,3 +55,66 @@ def __eq__(self, other):
+
+        Args:
+            subroutine (Schedule): A program to wrap with call instruction.
+        """
+        if name is None:
+            routine_id = hashlib.md5(str(subroutine.instructions).encode('utf-8')).hexdigest()
+        else:
+            routine_id = name
+
+        super().__init__((subroutine,), None,
+                         channels=tuple(subroutine.channels),
+                         name=routine_id)
+
```
Just an option. Not a strong suggestion
```suggestion
        """
        if name is None:
            name = hashlib.md5(str(subroutine.instructions).encode('utf-8')).hexdigest()

        super().__init__((subroutine,), None,
                         channels=tuple(subroutine.channels),
                         name=name)

```

lcapelluto reviewed (commented):
Nice work!

nkanazawa1989 commented on qiskit/pulse/transforms.py:
```diff
@@ -517,9 +521,29 @@ def align_func(schedule: Schedule,
+def flatten(program: ScheduleComponent) -> ScheduleComponent:
+    """Flatten any called nodes into a Schedule tree with no nested children.
+
+    ``Call`` instruction is also decomposed into raw instruction sequence.
```
Thanks for careful reading! I had forgotten to update docstring (I first included `remove_subroutines` in `flatten`...)

taalexander commented on releasenotes/notes/add-call-instruction-6f4fc37994090c2f.yaml:
```diff
@@ -0,0 +1,38 @@
+    The logic of the pulse builder is updated to store a call instruction when py:func:`pulse.call`
+    function is used with ``target``, i.e. ``QuantumCircuit`` or ``Schedule``.
+    Now this function saves the ``target`` as a call instruction rather than simply appending
+    its schedule to the context. This representation explicitly describes
+    the ``target`` is defined outside of the current scope, or as a subroutine.
+    No further schedule transforms are applied between subroutines and stored programs
+    are kept as-is. For example,
```
Same problems on confusion here.

taalexander commented on releasenotes/notes/add-call-instruction-6f4fc37994090c2f.yaml:
```diff
@@ -0,0 +1,38 @@
+
+    .. code-block:: python
+
+      xcirc = QuantumCircuit(1)
```
qc_x might be more canonical?

taalexander commented on releasenotes/notes/add-call-instruction-6f4fc37994090c2f.yaml:
```diff
@@ -0,0 +1,38 @@
+        pulse.call(xcirc)
+
+    the ``xcirc`` is called as an individual subroutine, thus these programs are not
+    merged within the bulder context thourgh consecutive x gates forms an identity.
```
```suggestion
    merged within the builder context through consecutive x gates forms an identity.
```

taalexander commented on releasenotes/notes/add-call-instruction-6f4fc37994090c2f.yaml:
```diff
@@ -0,0 +1,38 @@
+
+    the ``xcirc`` is called as an individual subroutine, thus these programs are not
+    merged within the bulder context thourgh consecutive x gates forms an identity.
+    The :py:class:`~qiskit.pulse.instructions.Call` instruction benefits a Qiskit compiler
+    to generate a cleaner target code. This instruction doesn't affect the underlying
+    program execution.
```
```suggestion
    The :py:class:`~qiskit.pulse.instructions.Call` instruction helps the Qiskit compiler
    to produce a cleaner and more reusable program representation. It does not affect the underlying
    program execution.
```

taalexander commented on test/python/pulse/test_builder.py:
```diff
@@ -462,46 +466,6 @@ def test_snapshot(self):
 
         self.assertEqual(schedule, reference)
 
-    def test_call_schedule(self):
```
I believe tests be kept until fully deprecated?

nkanazawa1989 replied:
not deprecated. this is moved to new unittest class which is a collection of tests involving the call instruction.

taalexander commented on qiskit/pulse/builder.py:
```diff
@@ -1617,12 +1666,21 @@ def call_circuit(circ: circuit.QuantumCircuit):
+    call(circ)
 
 
 def call(target: Union[circuit.QuantumCircuit, Schedule]):
```
Now that parameters are supported, I think it makes sense to be able to pass the parameters at call time here by name `pulse.call(sched, p1=0, ..., params=param_dict)`? Likewise, these should go in the `Call` instruction `Call(subroutine, consolidated_param_dict)`?

taalexander commented on qiskit/pulse/instructions/directives.py:
```diff
@@ -52,3 +55,66 @@ def __eq__(self, other):
                 set(self.channels) == set(other.channels))
+
+
+class Call(Directive):
```
This does not appear to be unit tested on its own anywhere.

taalexander reviewed (changes requested):
This looks like a great start! Excited to see this feature! 

I noticed that the `Call` Schedule instruction is not directly unit tested, this should be included.

eggerdj commented on qiskit/pulse/instructions/directives.py:
```diff
@@ -52,3 +55,66 @@ def __eq__(self, other):
+    def assign_parameters(self,
+                          value_dict: Dict[ParameterExpression, ParameterValueType]
+                          ) -> 'Call':
+        assigned_subroutine = self.subroutine.assign_parameters(value_dict)
```
Yes, I think this will mutate the called subroutine and I also think it will cause issues if the subroutine is called multiple times. I'm facing a similar issue in the calibration module.

nkanazawa1989 replied:
Good point. I'll update the logic.

eggerdj reviewed (changes requested):
Thomas raised some very good points.

eggerdj commented:
One extra question: what is the behavior if you deep copy a schedule with a call instruction? Do you also end up with a deep copy of the called schedule?

nkanazawa1989 commented on qiskit/visualization/pulse/matplotlib.py:
```diff
@@ -32,7 +32,11 @@
+                          Instruction, Schedule, ShiftFrequency, SetPhase)
+
+
+ScheduleComponent = Union[Schedule, Instruction]
+"""An element that composes a pulse schedule."""
```
Any suggestion of the place? I think defining this at schedule.py will cause cyclic import. So having this definition in __init__ of the top module is reasonable, but I'm not sure if you prefer this. Otherwise I will create types.py which has no dependency as I did for the visualization modules.

nkanazawa1989 commented on qiskit/pulse/instructions/directives.py:
```diff
@@ -52,3 +55,66 @@ def __eq__(self, other):
+        given these parameter values are managed by unique parameter objects.
+
+        Args:
+            subroutine (Schedule): A program to wrap with call instruction.
```
yes, I'll write todo for reminder

nkanazawa1989 commented on qiskit/pulse/builder.py:
```diff
@@ -417,35 +418,65 @@ def append_instruction(self, instruction: instructions.Instruction):
+            self._compile_lazy_circuit()
+            subroutine = self._compile_circuit(subroutine)
+
+        if len(subroutine.instructions) > 0:
+            self.append_instruction(instructions.Call(subroutine))
```
To cope with this unittest
https://github.com/Qiskit/qiskit-terra/blob/282e5334fae27114112a32cf99d9efcd83e1d8d9/test/python/pulse/test_builder.py#L216-L219

without this line, it generates an instruction with empty inline schedule. note that the circuit is transpiled with the setting defined in the context.

nkanazawa1989 commented on qiskit/pulse/builder.py:
```diff
@@ -417,35 +418,65 @@ def append_instruction(self, instruction: instructions.Instruction):
-            self.call_schedule(sched)
+            self.call_schedule(self._compile_circuit(lazy_circuit))
+
+    def _compile_circuit(self, circ) -> Schedule:
```
this helper does transpilation and schedule. it can incorporate some optimization passes which are currently not existing.

nkanazawa1989 commented:
Thanks for great feedback! The logic of parameter handling is update as follows:
- Call has parameter table that maps parameter object to assigned value. This is different from other instructions, i.e. in other instructions the table maps object to index of operand list.
- When parameter is assigned, the call instruction implicitly updates the parameter table, but no update is performed for subroutine. Thus actual parameter table in the subroutine is kept unbound.
- When `transforms.inline_subroutine` is called, this function creates a copy of subroutine (schedule) and assigns the parameter table of the target call instruction to copied subroutine.

This logic benefits two things:
- Remove overhead of deepcopy. Parameters are assigned only once when the main program is transformed.
- Without transforming, `Call` can keep the context of subroutine + parameter, i.e. function with arguments. Based on this data structure, we can create more sophisticated program representation for backend system.

The another point we need to discuss is the name of `Call` instruction. We may want to populate some database with a program involving call instruction. To generate correct foreign key for the database (assuming the subroutine is already defined in the database and we are calling the entry to generate a composite), this instruction need some instance variable to express that key. I can agree that generating hash for every call instruction is huge overhead. Perhaps we can define some subclass?

taalexander commented on qiskit/assembler/assemble_schedules.py:
```diff
@@ -20,16 +20,13 @@
 from qiskit.assembler.run_config import RunConfig
 from qiskit.exceptions import QiskitError
 from qiskit.pulse import instructions, transforms, library
+from qiskit.pulse.schedule import ScheduleComponentType
```
```suggestion
from qiskit.pulse.schedule import ScheduleComponent
```

taalexander commented on qiskit/pulse/builder.py:
```diff
@@ -1699,8 +1709,25 @@ def call(target: Union[circuit.QuantumCircuit, Schedule]):
+
+        with pulse.build() as main_prog:
+            pulse.call(subroutine, {amp: 0.1})
+            pulse.call(subroutine, {amp: 0.3})
```
Would it be possible to do
`pulse.call(subroutine, amp=0.1)`
or no because of the way parameters work?

taalexander commented on qiskit/pulse/instructions/call.py:
```diff
@@ -0,0 +1,129 @@
+
+    This instruction represents a call to a referenced pulse subroutine (schedule).
+    Note that this instruction is not exposed to end users.
+    The ``Call`` instruction is created with a target subroutine and it behaves as a
+    standard pulse instruction within the ``Schedule``. No physical operation to
+    a quantum processor is triggered by this instruction. On the other hand,
+    this instruction provides a set of inline function (subroutine) and
+    parameters to the compiler, and the compiler can use these information to generate
+    better program representation, especially, when the inline function is used multiple times
+    within the main program.
```
```suggestion
    The ``Call`` instruction represents the calling of a referenced subroutine (schedule).
    It enables code reuse both within the pulse representation and hardware (if supported).
```

taalexander commented on qiskit/pulse/instructions/call.py:
```diff
@@ -0,0 +1,129 @@
+    within the main program.
+    """
+    # Prefix to use for auto naming.
+    prefix = 'subroutine-'
```
```suggestion
    prefix = 'call'
```

taalexander commented on qiskit/pulse/instructions/call.py:
```diff
@@ -0,0 +1,129 @@
+                the hash of instructions of the subroutine.
+        """
+        if name is None:
+            name = self.prefix + subroutine.name
```
```suggestion
            name = f"{self.prefix}_{subroutine.name}"
```

taalexander commented on qiskit/pulse/instructions/call.py:
```diff
@@ -0,0 +1,129 @@
+    # Prefix to use for auto naming.
+    prefix = 'subroutine-'
+
+    def __init__(self, subroutine, name: Optional[str] = None):
```
I believe we should encourage parameters to be assigned at instantiation time for the `Call` instruction. This is a container for a schedule and a set of parameters that will later be bound to it. We should not encourage this class to be mutable. eg. something like,
```suggestion
    def __init__(self, subroutine, value_dict: Dict[param, any], name: Optional[str] = None):
```

nkanazawa1989 replied:
I added value dict to the constructor, but still this instruction is mutable. We may want to define a schedule template by calling another subroutine which is parameterized, because parameters may drift from time to time. 

For example, writing this for every subroutine call is bit annoying.
```python
pulse.call(subroutine, amp=subroutine.get_parameters('amp')[0], sigma=subroutine.get_parameters('sigma')[0], ...)
```
So I'd like to keep an option that enable us to insert subroutine without parameter specification.

taalexander commented on qiskit/pulse/instructions/call.py:
```diff
@@ -0,0 +1,129 @@
+    @property
+    def duration(self) -> Union[int, ParameterExpression]:
+        """Duration of this instruction."""
+        return self.operands[0].duration
```
```suggestion
        return self.subroutine.duration
```

taalexander commented on qiskit/pulse/transforms.py:
```diff
@@ -531,13 +528,30 @@ def flatten(program: ScheduleComponent) -> ScheduleComponent:
-def remove_subroutines(program: Schedule):
-    """Recursively removes the call instruction."""
+def inline_subroutines(program: Schedule) -> Schedule:
+    """Recursively remove the call instruction and inlines the attached instructions.
```
```suggestion
    """Recursively remove call instructions and inline the respective subroutine instructions.
```

taalexander commented on releasenotes/notes/add-call-instruction-6f4fc37994090c2f.yaml:
```diff
@@ -13,26 +13,41 @@ deprecations:
-    No further schedule transforms are applied between subroutines and stored programs
-    are kept as-is. For example,
+    This instruction represents calling a schedule as a subroutine from within another schedule.
+    This is not exposed to end users, but internally used for compiling purpose.
```
Is this really not exposed to end-users? It seems like any user of a Schedule could use this straightforwardly?

eggerdj replied:
Are we sure we do not want to expose this to the user? It could be very useful in the calibration module for instance.

nkanazawa1989 replied:
I think this depends on how we define _expose_. I don't think a developer of calibration module is not an end user. Because Call instruction doesn't trigger any operation on the Schedule so user should not aware of its existence. This is just a compiler issue and we can run the same program without Call. I'm not sure if we really need to ask end users to program the pulse schedule by considering compiler stuff, i.e.

I think this is quite easy to understand
```python
xsched = Schedule()
xsched += Play(Drag(160, 0.1, 40, 0.5), DriveChannel(0))

sched = Schedule()
sched += xsched
sched += Play(Drag(160, 0.2, 40, 0.5), DriveChannel(0))
```

but I think below example is very confusing if a user is not a software engineer (though Call is quite popular instruction which can be seen in many programming languages):
```python
sched = Schedule()
sched += Call(xsched)
sched += Play(Drag(160, 0.2, 40, 0.5), DriveChannel(0))  # why this is not Call ed? - of course it can be
```

Note that above examples will generate identical pulse sequence, and thus user may confuse when to insert the Call instruction. Actually the user can use Call for every `+=` sugar, but this will just mess up the target code.

In the pulse builder this is more intuitive because it has notion of context
```python
with pulse.build() as sched:
    pulse.call(xsched)
    pulse.play(Drag(160, 0.2, 40, 0.5), DriveChannel(0))
```
It can be understood Call is applied if we reuse schedules defined outside of the current context.

nkanazawa1989 replied:
I think this is one of reasons we want to discourage direct use of Schedule.

taalexander commented on releasenotes/notes/add-call-instruction-6f4fc37994090c2f.yaml:
```diff
@@ -13,26 +13,41 @@ deprecations:
+      amp = Parameter('amp')
+
+      with pulse.build() as subroutine:
+        pulse.play(pulse.Gaussian(160, amp, 40), DriveChannel(0))
+
+      with pulse.build() as main_prog:
+        pulse.call(subroutine, {amp: 0.1})
+        pulse.call(subroutine, {amp: 0.3})
```
Tabbing seems like only 2 spaces here.

taalexander reviewed (changes requested):
Looks very good. Some relatively minor comments in this round.

eggerdj commented on qiskit/pulse/instructions/call.py:
```diff
@@ -0,0 +1,129 @@
+
+        Parameter values are not immediately assigned. The subroutine with parameters
+        assigned according to the populated parameter table will be generated only when
+        :func:`~qiskit.pulse.transforms.inline_subroutines` function is applied to this
+        instruction. Note that parameter assignment logic creates a copy of subroutine
```
When is this typically called?

eggerdj commented on qiskit/pulse/library/waveform.py:
```diff
@@ -97,6 +97,11 @@ def _clip(self, samples: np.ndarray, epsilon: float = 1e-7) -> np.ndarray:
+    @property
+    def parameters(self) -> Dict[str, Any]:
+        """Return a dictionary containing the pulse's parameters."""
+        return dict()
```
Why do we return an empty dict?

nkanazawa1989 replied:
Because waveform doesn't have parameters. Actually this property doesn't exist before, but I need to define this to equally handle `ParametricPulse` and `Waveform` objects in some logic.

eggerdj commented on qiskit/pulse/builder.py:
```diff
@@ -416,36 +418,74 @@ def append_instruction(self, instruction: instructions.Instruction):
+            self._compile_lazy_circuit()
+            subroutine = self._compile_circuit(subroutine)
+
+        if len(subroutine.instructions) > 0:
+            call_def = instructions.Call(subroutine)
+            if value_dict:
+                call_def.assign_parameters(value_dict)
+            self.append_instruction(call_def)
```
Is this problem solved using the `inline_subroutines` function?

nkanazawa1989 replied:
yes, it's solved by the inline function. `assign_parameters` logic is overwritten and Call instruction itself doesn't update attached subroutine. It just stores the parameter dictionary separately.

eggerdj commented on qiskit/pulse/instructions/play.py:
```diff
@@ -75,6 +75,22 @@ def duration(self) -> Union[int, ParameterExpression]:
         """Duration of this instruction."""
         return self.pulse.duration
 
+    def _initialize_parameter_table(self,
+                                    operands: Tuple[Any]):
```
Why does this function have different implementations for different instructions?

nkanazawa1989 replied:
Because parameters are stored in different place and the parameter table usually defines a mapping of parameter object to where it's stored, i.e. index of operand tuple.

For example, shift phase
```
operands = (phase, channel)
```
When we parameterize phase and channel, this will generate a table `{phase: [0], channel: [1]}`. This is the standard logic to create the table.

However, in Call instruction
```
operands = (subroutine, )
```
and the `subroutine` cannot be `ParameterExpression` object. If we use the same logic, the table always becomes `{amp: [0], phase: [0], ...}`. Since value is not stored in the parameter table, we need to assign values immediately with object copy. This is huge overhead. Instead, I define a parameter table as something like `{amp: 0.1, phase: 0.5}`, i.e. a mapping of object to assigned value. This is more convenient because we can later assign parameters in `inline_subroutine` function with this table. Object copy is not performed at every parameter assignment.

eggerdj replied:
Thanks.

eggerdj reviewed (changes requested):
Looking good. A few points to improve on from my side.

nkanazawa1989 commented:
Thanks, I updated docstrings accordingly, and updated the logic of call function. Now we can handle parameters as kwargs at the builder level, i.e.

```python
with pulse.build() as sched:
   pulse.call(subroutine, amp=0.1, sigma=40, ...)
```

taalexander commented on qiskit/pulse/builder.py:
```diff
@@ -451,28 +450,37 @@ def call_schedule(self, schedule: Schedule):
-                        ):
-        """Call a schedule defined outside of the current scope.
+                        name: Optional[str] = None,
+                        **kw_params):
```
```suggestion
                        **kw_params,
                        value_dict=None):
```
For consistency, we should also support `value_dict` for cases many parameters may share a name.

taalexander commented on qiskit/pulse/builder.py:
```diff
@@ -451,28 +450,37 @@ def call_schedule(self, schedule: Schedule):
+                    for param_obj in param_objs:
+                        value_dict[param_obj] = assigned_value
+                else:
+                    warnings.warn(f'Parameter {param_name} is not defined in the '
```
Maybe this should just be an error?

nkanazawa1989 commented:
@taalexander Thanks! I updated parameter handling based on your suggestion.

taalexander commented on qiskit/pulse/builder.py:
```diff
@@ -451,7 +452,8 @@ def call_schedule(self, schedule: Schedule):
                         subroutine: Union[circuit.QuantumCircuit, Schedule],
                         name: Optional[str] = None,
-                        **kw_params):
+                        value_dict: Optional[Dict[ParameterExpression, ParameterValueType]] = None,
+                        **kw_params: ParameterValueType):
```
I wonder if these should be swapped? Should we force the user to specify `call(routine, value_dict={...})`, I lean towards yes as in the future if there was a canonical ordering for args we could do `call(routine, arg0, ..., argn=val)` with both args and kwargs. Do you have thoughts on this?

nkanazawa1989 replied:
You mean adding `*args: ParameterValeType` ? Could you please elaborate on _swapped_? I think we cannot define keyword augment (`kw_params`) before the positional argument (`value_dict`) and `ParameterExpression` object cannot be a keyword. I think adding `*args` is good idea. However `Schedule.parameters` is a `Set` object and I don't think the order is preserved.

taalexander replied:
Resolving in favour of adding the notion of a function in a future PR.
